File: packages/agent-server/src/lib/chat-select.ts
/**
 * Programmatic chat selection via a11y-guided click.
 *
 * Uses the chat-select tool to open a specific chat in the WeChat UI.
 */

import { execCommand, type ExecOptions } from "./exec.js";

interface OpenChatResult {
  ok: boolean;
  username?: string;
  index?: number;
  error?: string;
}

interface ListSessionsResult {
  ok: boolean;
  sessions?: Record<string, number>;
  error?: string;
}

/**
 * Open a chat in the WeChat UI by username.
 *
 * This triggers WeChat to select the chat, which downloads pending media
 * and clears unread counts.
 */
export async function openChat(
  chatId: string,
  options?: ExecOptions
): Promise<OpenChatResult> {
  const result = await execCommand("chat-select", [chatId], {
    ...options,
    timeout: 120_000, // Chat selection can be slow
  });

  // Log stderr (debug output from chat-select.py)
  if (result.stderr) {
    for (const line of result.stderr.split("\n")) {
      if (line.trim()) console.log(line);
    }
  }

  try {
    return JSON.parse(result.stdout);
  } catch {
    return {
      ok: false,
      error: result.stderr || result.stdout || `chat-select exited with code ${result.exitCode}`,
    };
  }
}

/**
 * List all sessions visible in WeChat's session vector.
 *
 * Returns a map of username -> index in the session list.
 */
export async function listChatSessions(
  options?: ExecOptions
): Promise<ListSessionsResult> {
  const result = await execCommand("chat-select", ["--list"], {
    ...options,
    timeout: 120_000,
  });

  try {
    return JSON.parse(result.stdout);
  } catch {
    return {
      ok: false,
      error: result.stderr || result.stdout || `chat-select exited with code ${result.exitCode}`,
    };
  }
}

File: packages/agent-server-rust/src/tools/chat_select.rs
use super::exec::{exec_command, ExecOptions};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenChatResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skipped: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Open a chat in the WeChat UI using the chat-select tool.
pub async fn open_chat(
    chat_id: &str,
    force: bool,
    click_xy: Option<(f64, f64)>,
) -> OpenChatResult {
    let mut args = vec!["--chat-id", chat_id];

    let xy_str;
    if let Some((x, y)) = click_xy {
        xy_str = format!("{},{}", x as i32, y as i32);
        args.push("--click-xy");
        args.push(&xy_str);
    }

    if force {
        args.push("--force");
    }

    let args_str: Vec<&str> = args.iter().copied().collect();

    let result = exec_command(
        "python3",
        &[&["/opt/tools/chat-select.py"], args_str.as_slice()].concat(),
        &ExecOptions::default(),
    )
    .await;

    if result.exit_code != 0 {
        return OpenChatResult {
            ok: false,
            skipped: None,
            error: Some(result.stderr),
        };
    }

    // Parse JSON output
    serde_json::from_str(&result.stdout).unwrap_or(OpenChatResult {
        ok: true,
        skipped: None,
        error: None,
    })
}
